lib/Mongo/MongoDB.php

```diff
@@ -405,7 +405,12 @@ public function command(array $data, $options = [], &$hash = null)
     {
         try {
             $cursor = new \MongoCommandCursor($this->connection, $this->name, $data);
-            $cursor->setReadPreference($this->getReadPreference());
+
+            if (!empty($data['new']) || !empty($data['upsert'])) {
```

Comment:
https://github.com/alcaeus/mongo-php-adapter/issues/181 and https://github.com/doctrine/mongodb/pull/297 should address the issue. The adapter was missing some behavior that forced a primary server for most commands, which was undocumented behavior in the legacy `mongo` extension. On a separate note, Doctrine MongoDB needed a patch to be able to apply a read preference to `findAndModify` and `mapReduce` commands executed through its query builder.